[wip] enhanced relational comparisons

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

fixes #21429

cf #9154

#### Brief description of what is fixed or changed

Simplification of Relationals with denominators of known sign is better handled now so

```python
>>> x = var('x',positive=1)
>>> 1/x < 1
1/x < 1
>>> simplify(_)
x > 1
```

In addition, `floor` and `ceiling` now are aware of arguments of size with magnitude larger or smaller than 1:
```python
floor(-1/Dummy(prime=True)) == -1
```
#### Other comments

This picture is useful for checking assumptions that have been coded.

![image](https://user-images.githubusercontent.com/90703/119186140-464aa300-ba3d-11eb-8e85-b01435248c7a.png)

What has been difficult is knowing what cases to code in the `_eval_is_ge` routines since some expressions will evaluate with assumptions if these handlers do nothing. So I have tried to add every case. I have also attempted to not do any recalculation of any expression other than to multiple or add Numbers to existing Numbers.

- [ ] fix this

```python
    >>> Dummy(zero=True)**Dummy(infinite=True) >= 0
    True  <-- should be None
    >>> Dummy(zero=True)**Dummy(infinite=True) > 0
    False  <-- should be None
    >>> 0**-oo
    zoo
```
#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * floor and ceiling recognize unit fractions whose magnitude is less than 1
* core
  * better simplification of Rationals with rational expressions
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/functions/elementary/integers.py:
```diff
@@ -20,11 +20,24 @@ class RoundFunction(Function):
         if v is not None:
             return v
 
+        n, d = fraction(arg)
+        if d.is_integer:
+            if cls is floor:
+                if n == -1 and d.is_positive:
+                    return n
+                if n == 1 and d.is_negative:
+                    return -n
+            if cls is ceiling:
+                if n == 1 and d.is_positive:
+                    return n
+                if n == -1 and d.is_negative:
+                    return -n
```
It seems odd to add this as a special case here. Really what this block is doing is:
```python
if arg.is_nonnegative and is_lt(arg, S.One):
    return S.Zero
elif arg.is_negative and is_ge(arg, S.NegativeOne):
    return S.NegativeOne
```
We could make `is_ge/is_lt` handle this by including handlers for `_eval_is_ge(Pow, ...)` and so on.

To me it would seem more useful for the codebase as a whole to have the code here depend on general purpose handlers and to improve the capabilities of the general purpose handlers for use in other parts of the codebase as well.

oscarbenjamin replied:
Note that adding this capability to `is_ge` would immediately make it available to `Relational` as well.

smichr commented:
In `relational.py` I define
```
@dispatch(Pow, Expr)
def _eval_is_le(lhs, rhs):
```
but the following does not dispatch there:
```python
>>> from sympy.core.relational import is_le
>>> var('i',integer=1,positive=1)
>>> is_le(1/i,S.One)
```

oscarbenjamin commented:
Everything is defined in terms of `_eval_is_ge` so there is no `_eval_is_le`. In your case
```
is_le(1/i, 1)   -->  is_ge(1, 1/i)
```
So what is needed is 
```
@dispatch(Expr, Pow)
def _eval_is_ge(lhs, rhs):
```
Or perhaps something more precise like:
```
@dispatch(type(S.One), Pow)
def _eval_is_ge(lhs, rhs):
    base, exp = rhs.as_base_exp()
    if exp.is_nonpositive and base.is_positive and is_le(base, S.One):
        return True 
```

It looks like `_eval_is_ge` is barely developed so far. There was some work on it but it stalled https://github.com/sympy/sympy/pull/19948

smichr commented on sympy/functions/elementary/integers.py:
```diff
@@ -20,11 +20,23 @@ class RoundFunction(Function):
     @classmethod
     def eval(cls, arg):
-        from sympy import im
+        from sympy import im, fraction
```
```suggestion
        from sympy import im
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
+            return True
+
+
+@dispatch(Number, Mul)
```
```suggestion
@dispatch(Number, Mul) # type: ignore
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
             return dif
 
 
+@dispatch(Number, Pow)
```
```suggestion
@dispatch(Number, Pow) # type: ignore
```

smichr commented:
> Everything is defined in terms of `_eval_is_ge`

I get the idea, but this is not an idiom I grew up with. Having worked on the assumption system, this addition is a welcome one. I think I have this working. The tests I added and the floor/ceiling tests passed. If the rest pass, this is ready for a review, @oscarbenjamin .

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
             return dif
 
 
+@dispatch(Number, Pow) # type: ignore
+def _eval_is_ge(lhs, rhs):
```
```suggestion
@dispatch(Number, Pow)
def _eval_is_ge(lhs, rhs): # noqa:F811
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
+            return True
+
+
+@dispatch(Pow, Number) # type: ignore
+def _eval_is_ge(lhs, rhs):
```
```suggestion
@dispatch(Pow, Number)
def _eval_is_ge(lhs, rhs): # noqa:F811
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
+            return True
+
+
+@dispatch(Number, Mul) # type: ignore
+def _eval_is_ge(lhs, rhs):
```
```suggestion
@dispatch(Number, Mul)
def _eval_is_ge(lhs, rhs): # noqa:F811
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
+            return True
+
+
+@dispatch(Mul, Number)  # type: ignore
+def _eval_is_ge(lhs, rhs):
```
```suggestion
@dispatch(Mul, Number)
def _eval_is_ge(lhs, rhs): # noqa:F811
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,6 +1087,40 @@ def _n2(a, b):
+
+
 @dispatch(Expr, Expr)
 def _eval_is_ge(lhs, rhs):
```
```suggestion
def _eval_is_ge(lhs, rhs): # noqa:F811
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,8 +1087,42 @@ def _n2(a, b):
+def _eval_is_ge(lhs, rhs): # noqa:F811
+    from sympy.simplify.radsimp import fraction
+    num, den = fraction(rhs)
+    if num.is_Number and den != 1:
+        if is_le(den, lhs/num):
+            return True
```
if num and num.is_Number and den != 1:
    a, b= den, lhs/num
    if (num < 0):
        a, b = b, a
    if is_le(a, b):
        return True

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,8 +1087,42 @@ def _n2(a, b):
+def _eval_is_ge(lhs, rhs): # noqa:F811
+    from sympy.simplify.radsimp import fraction
+    num, den = fraction(lhs)
+    if num.is_Number and den != 1:
+        if is_ge(den, rhs/num):
+            return True
```
```suggestion
```suggestion
    if num and num.is_Number and den != 1:
        a, b = den, rhs/num
        if (num < 0):
            a, b = b, a
        if is_ge(a, b):
            return True
``````

smichr commented on sympy/core/relational.py:
````diff
@@ -1075,8 +1087,49 @@ def _n2(a, b):
+def _eval_is_ge(lhs, rhs): # noqa:F811
+    from sympy.simplify.radsimp import fraction
+    num, den = fraction(lhs)
+```suggestion
````
```suggestion
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,8 +1087,48 @@ def _n2(a, b):
             return dif
 
 
+@dispatch(Number, Pow)
+def _eval_is_ge(lhs, rhs): # noqa:F811
+    if lhs:
+        base, exp = rhs.as_base_exp()
+        if exp.is_nonpositive and is_ge(base, 1/lhs):
+            return True
```
We should try to avoid creating new Basic instances like `1/lhs`.

Is the logic here correct?

If we have
```
1/2 >= 5 ** 0
```
Then
```
lhs = 1/2
base = 5
exp = 0
```
So `exp.is_nonpositive = True` and `base >= 1/lhs` so we return True but the inequality is not true.

smichr replied:
I think that is why the tests are failing...I have to look more closely at this. 

As to creating different objects, that is the purpose of working with only Number but perhaps I should have limited that to Rational and then we wouldn't be making a new type.

smichr commented:
working on less evaluation

smichr commented:
@oscarbenjamin , if you get a chance would you please take a look again. It's not really clear to me what is supposed to be handled where. So there are many other cases that could *potentially* be handled in the Rational/Pow cases but many of them evaluate before getting to these `_eval_is_ge` routines.

And I added Expr/Pow handlers but was warned about ambiguity so I just added the handling directly in the Expr/Expr stub.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1087,72 @@ def _n2(a, b):
+        return
+    base, exp = rhs.as_base_exp()
+    if base.is_nonnegative and (exp + 1).is_nonpositive:
+        return is_ge(base, rat)
```
Here `exp + 1` creates a new Basic instance. It would be better to use `is_ge(exp, 1)`.

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1087,72 @@ def _n2(a, b):
 
+@dispatch(Rational, Pow)
+def _eval_is_ge(rat, rhs): # noqa:F811
+    if rat.is_zero:
+        return
+    base, exp = rhs.as_base_exp()
+    if base.is_nonnegative and (exp + 1).is_nonpositive:
```
```suggestion
    # i >= b**((+/-1)/j)
    if rat.is_zero:
        return
    base, exp = rhs.as_base_exp()
    if base.is_nonnegative and is_le(exp, S.NegativeOne):
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1087,79 @@ def _n2(a, b):
+    base, exp = lhs.as_base_exp()
+    if base.is_nonnegative:
+        if is_ge(base, S.One) and is_ge(exp, S.One):
+            return is_le(rat, S.One)
```
So we want to know if `a**b >= c`. The inference here is
```
a >= 1  &  b >= 1    --->     a**b >= c  iff  c <= 1
```
That's not an if and only if though because e.g. if a = 2 and b = 2 and c = 2 then this would return False even though the statement is true.

smichr replied:
```suggestion
            if is_le(rat, S.One):
                return True
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1087,79 @@ def _n2(a, b):
+    base, exp = rhs.as_base_exp()
+    if base.is_nonnegative:
+        if is_le(exp, S.NegativeOne):
+            return is_ge(base, rat)
```
```suggestion
            return is_ge(base, 1/rat)  # rat >= b^(e<=-1) -> b^(e>=1) >= 1/rat
```

smichr commented on sympy/core/relational.py:
```diff
@@ -6,11 +6,15 @@
 from .basic import Basic
 from .evalf import EvalfMixin
 from .function import AppliedUndef
+from .mul import Mul
```
```suggestion
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1087,93 @@ def _n2(a, b):
-def _eval_is_ge(lhs, rhs):
-    return None
+def _eval_is_ge(lhs, rhs): # noqa:F811
+    # simplify expression in case one side is Rational
```
```suggestion
    from sympy.simplify.radsimp import fraction
    # simplify expression in case one side is Rational
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1086,94 @@ def _n2(a, b):
+    op = is_le, is_ge
+    ix = -1
+    L, R = lhs, rhs
+    if L.is_Rational:
```
```suggestion
    if L.is_Rational:
        if R.is_Rational:
            return  # normal assumptions will handle this
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1086,96 @@ def _n2(a, b):
+    if base.is_nonnegative:
+        if is_ge(base, S.One) and is_ge(exp, S.One):
+            if is_le(rat, S.One):
+                return True
```
So this is `a ** b >= c` and we check `a >= 0` and then `a >= 1` and `b >=1` then if `c <= 1` we return True. I think we only need `b >= 0` here.

Essentially this is:
```
def is_ge(lhs, rat):
    if is_ge(lhs, S.One) and is_le(rat, One):
        return True
```
Maybe it makes sense to write it like that and let a `Pow`, `One` handler take care of further details. Perhaps one of the most important usages of the `is_ge` framework is comparisons against `S.One`...

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1086,92 @@ def _n2(a, b):
+    L, R = lhs, rhs
+    if L.is_Rational:
+        L, R = R, L
+        ix = ~ix
+    if R.is_Rational:
+        c, t = L.as_coeff_Add()
+        if c:
+            L, R = t, R - c
+        n, d = fraction(L)
+        c, n = n.as_coeff_Mul()
+        dc, d = d.as_coeff_Mul()
+        c /= dc  # ratio of Rational -> Rational
+        if d is S.One and c is not S.One:
+            # join Rationals c and R
+            if c < 0:
+                ix = ~ix
+            if is_ge(d, S.Zero):
+                return op[ix](n, R/c)
+            if is_le(d, S.Zero):
+                return op[~ix](n, R/c)
+        elif d is not S.One:
+            # bring denom to numerator on other side
+            # c*n/d ~ R -> c*n/R ~ d, watching for signs of R and d
+            if R < 0:
+                ix = ~ix
+            if is_ge(d, S.Zero):
+                return op[ix](c*n/R, d)
+            if is_le(d, S.Zero):
+                return op[~ix](c*n/R, d)
+
+    # x >= x**e
+    if isinstance(rhs, Pow) and not isinstance(lhs, Pow):
+        if rhs.base == lhs and lhs.is_nonnegative and is_le(
+                rhs.exp, S.One):
+            return True
+    # x**e >= x
+    if isinstance(lhs, Pow) and not isinstance(rhs, Pow):
+        if lhs.base == rhs and rhs.is_nonnegative and is_ge(
+                lhs.exp, S.One):
+            return True
```
```suggestion
def _eval_is_ge(lhs, rhs): # noqa:F811
    from sympy.core.function import expand_mul

    if lhs.is_Atom and rhs.is_Atom:
        return  # resolve with assumptions of lhs, rhs

    # x >= x**e
    if isinstance(rhs, Pow) and not isinstance(lhs, Pow):
        if rhs.base == lhs and lhs.is_nonnegative and is_le(
                R.exp, S.One):
            return True
    # x**e >= x
    if isinstance(lhs, Pow) and not isinstance(rhs, Pow):
        if lhs.base == rhs and R.is_nonnegative and is_ge(
                lhs.exp, S.One):
            return True

    # make expression canonical wrt Rationals
    # to minimize the number of handlers needed, e.g.
    # -1/x >= 2 is Mul, Integer but can be Pow, Integer by
    # rewriting as 1/x <= -2
    op = is_le, is_ge
    ix = -1
    L, R = lhs, rhs
    n, d = [expand_mul(i) for i in (L - R).as_numer_denom()]
    if is_ge(d, S.Zero):
        L = n
    elif is_le(d, S.Zero):
        L = n
        ix = ~ix
    else:
        L = n/d
    a, L = L.as_coeff_Add()
    R = -a
    m, L = L.as_coeff_Mul()
    if m < 0:
        ix = ~ix
    R /= m
    if (L, R) != (lhs, rhs):
        return op[ix](L, R)
```

smichr commented on sympy/core/tests/test_relational.py:
```diff
@@ -1211,3 +1226,48 @@ def _eval_is_ge(lhs, rhs):
+    assert i**pi >=0
+    j = Symbol('j', integer=True, nonnegative=True)
+    r = i/(i + j)
+    assert r <= 1
```
```suggestion
    assert r <= 1
    i, j = symbols('i j', integer=True, positive=True)
    assert 1/i +1 <= j + 1
    assert (1/i + 1)/(j + 1) <= 1

```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1086,82 @@ def _n2(a, b):
+    # x >= x**e
+    if isinstance(rhs, Pow) and not isinstance(lhs, Pow):
+        if rhs.base == lhs and lhs.is_nonnegative and is_le(
+                R.exp, S.One):
```
```suggestion
                rhs.exp, S.One):
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1086,82 @@ def _n2(a, b):
+            return True
+    # x**e >= x
+    if isinstance(lhs, Pow) and not isinstance(rhs, Pow):
+        if lhs.base == rhs and R.is_nonnegative and is_ge(
```
```suggestion
        if lhs.base == rhs and rhs.is_nonnegative and is_ge(
```

smichr commented on sympy/core/relational.py:
```diff
@@ -327,6 +330,14 @@ def _eval_simplify(self, **kwargs):
                     from sympy.solvers.solveset import linear_coeffs
                     x = free.pop()
                     dif = r.lhs - r.rhs
+                    # r.func(dif, 0) ->
+                    # r.func(n/d, 0) ->
+                    # r.func(sign(d)*n, 0)
+                    n, d = dif.as_numer_denom()
+                    if d.is_positive:
+                        dif = n
+                    elif d.is_negative:
+                        dif = -n
```
```suggestion
```

smichr commented on sympy/core/tests/test_relational.py:
```diff
@@ -787,6 +787,18 @@ def test_simplify_relational():
     assert Lt(x, 2).simplify() == Lt(x, 2)
     assert Lt(-x, 2).simplify() == Gt(x, -2)
 
+    p = symbols('n', integer=True, positive=True)
+    assert (1/p <= 1).simplify() == S.true
+    assert (-1/p >= -1).simplify() == S.true
+    n = symbols('n', integer=True, negative=True)
+    assert (1/n >= -1).simplify() == S.true
+    assert (-1/n <= 1).simplify() == S.true
+    p = symbols('p', prime=True)
+    assert p/n <= 0
+    n = symbols('n', negative=True)
+    assert (1/n < 1/(n - 1)).simplify() == S.true
+    assert ((n + 1)/(n - 1) > 1).simplify() == S.false
+
```
```suggestion
```

smichr commented on sympy/core/tests/test_relational.py:
```diff
@@ -1211,3 +1226,51 @@ def _eval_is_ge(lhs, rhs):
+    assert r <= 1
+    i, j = symbols('i j', integer=True, positive=True)
+    assert 1/i +1 <= j + 1
+    assert (1/i + 1)/(j + 1) <= 1
```
```suggestion
    assert (1/i + 1)/(j + 1) <= 1
    n = symbols('n', negative=True)
    assert 1/n < 1/(n - 1)
    assert ((n + 1)/(n - 1) > 1) == S.false
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,184 @@ def _n2(a, b):
+            ix = ~ix
+        R /= c
+    if R.is_Rational:
+        was = L,R
```
```suggestion
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,183 @@ def _n2(a, b):
-    return None
+def _eval_is_ge(lhs, rhs):  # noqa:F811
+    from sympy import Mod, fraction
+    if lhs.is_Atom and rhs.is_Atom:
```
```suggestion
    if lhs.is_Atom and rhs.is_Atom or (
            lhs.is_number and rhs.is_number):
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,184 @@ def _n2(a, b):
+            elif is_le(d, S.Zero):
+                ix = ~ix
+                L, R = n/r, den
+    if (L, R) != (lhs, rhs):
```
```suggestion
    if {L, R} != {lhs, rhs}:
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,184 @@ def _n2(a, b):
+            elif is_le(d, S.Zero):
+                ix = ~ix
+                L, R = n/r, den
+    if {L, R} != {lhs, rhs}:
```
```suggestion
    if {L, R} - {lhs, rhs, -lhs, -rhs}:
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,182 @@ def _n2(a, b):
+def _eval_is_ge(lhs, rhs):  # noqa:F811
+    from sympy import Mod, fraction
+    if lhs.is_Atom and rhs.is_Atom or (
+            lhs.is_number and rhs.is_number):
```
```suggestion
            lhs.is_number and rhs.is_number or
            not (lhs.is_finite and rhs.is_finite)):
```

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -962,7 +962,7 @@ def reduce_inequalities(inequalities, symbols=[]):
 
     >>> reduce_inequalities(0 <= x + y*2 - 1, [x])
-    (x < oo) & (x >= 1 - 2*y)
+    x >= 1 - 2*y
```
```suggestion
    (x < oo) & (x >= 1 - 2*y)
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,220 @@ def _n2(a, b):
+        R = mr/m*r + (br - b)/m
+        if m < 0:
+            ix = ~ix
+    if R.is_Rational:
```
```suggestion
    if R.is_Number:
```

smichr commented:
@oscarbenjamin I will start to add tests specifically to target all uncovered lines but I think the general structure is ready for review. I will clean up the commit history before committing once this has been checked over and iterated to completion.

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,220 @@ def _n2(a, b):
+        if r is S.Zero:
+            return is_ge(S.Zero, b)
+        # r == -1
+        if b is S.NegativeOne:
```
```suggestion
        if b is S.NegativeOne and is_ge(e, S.Zero):
```

smichr commented on sympy/core/relational.py:
```diff
@@ -6,11 +6,14 @@
 from .basic import Basic
 from .evalf import EvalfMixin
 from .function import AppliedUndef
+from .numbers import Rational
```
```suggestion
from .mod import Mod
from .numbers import Rational
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,220 @@ def _n2(a, b):
+
+@dispatch(Pow, Rational) # type: ignore
+def _eval_is_ge(_p, r):  # :F811
+    from sympy import Mod
```
```suggestion
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,220 @@ def _n2(a, b):
+
+@dispatch(Rational, Pow) # type: ignore
+def _eval_is_ge(r, _p):  # noqa:F811
+    from sympy import Mod
```
```suggestion
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1078,220 @@ def _n2(a, b):
-def _eval_is_ge(lhs, rhs):
-    return None
+def _eval_is_ge(lhs, rhs):  # noqa:F811
+    from sympy import Mod, fraction
```
```suggestion
    from sympy.simplify.radsimp import fraction
```

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,218 @@ def _n2(a, b):
+                ix = ~ix
+                L, R = n/r, den
+    # if we did more than change sign or switch sides, re-evaluate
+    if {L, R} - {lhs, rhs, -lhs, -rhs}:
```
it seems to me that the return should be something like `is_ge(L,R)` if either is a power else `L >= R` if L and R are not trivially the same as at the start.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,218 @@ def _n2(a, b):
+@dispatch(Pow, Rational) # type: ignore
+def _eval_is_ge(_p, r):  # :F811
+    if _p.is_number:
+        return
```
Is this return intended to allow the floating point evaluation to handle this case?

Part of the idea of introducing `is_ge` et al was to replace/eliminate the use of numerical evaluation in these queries so that they can be resolved symbolically and therefore be able to handle arbitrarily large numbers.

smichr replied:
I think I have tightened up the detection of infinite recursion perhaps I can relax this constraint.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,218 @@ def _n2(a, b):
+    if _p.is_number:
+        return
+    if r <= -1:
+        return is_ge(S.NegativeOne, _invpow(_p))
```
Here we are testing `_p >= r` where `_p = a**b` so `a**b >= r`

Now if `r <= -1` we return `-1 >= a**-b`.

If `a = 2` and `b = 3` and `r = -2` then for `2**3 >= -2` (which is true) we return `-1 >= 2**-3` (which is false).

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,267 @@ def _n2(a, b):
+                l = S.Zero
+            else:
+                return
+        elif l.base.is_zero is None:
```
```suggestion
        elif l.base.is_nonnegative is None:
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,281 @@ def _n2(a, b):
+_small_neg = lambda x: fuzzy_and((
+    is_ge(x, S.NegativeOne), is_ge(S.Zero, x)))
+
+_big_ppow = lambda b, e: fuzzy_or((
+        fuzzy_and((b.is_zero, e.is_zero)),  # 1
+        fuzzy_and((Mod(e, 2) == 0, fuzzy_or((
+        is_ge(S.NegativeOne, b) and is_ge(e, S.Zero),
+        _small_neg(b) and is_ge(S.Zero, e))))),
+    _small_pos(b) and is_ge(S.Zero, e),
+    is_ge(b, S.One) and is_ge(e, S.Zero)))
+
```
This kind of thing is definitely better as a proper function with a docstring rather than a lambda function. If the function is internal then it does not need to have a "full" docstring but at least an explanation of what it tests. I don't know what `_big_ppow` means and it would take a long time to parse it from the code here!

smichr replied:
I'm in the mid-game right now. I really hope this can be a lot simpler.

smichr replied:
for the record, the helpers identify whether a power is positive or negative and big (>= 1) or small (between 0 and 1).

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1079,281 @@ def _n2(a, b):
+
+    if L.is_Pow or R.is_Pow:
+        b, e = L.as_base_exp()
+        B, E = R.as_base_exp()
```
The idea with multipledispatch is supposed to be that we can write a `Pow/Pow` handler rather than having all the logic in an `Expr/Expr` handler that checks the types.

smichr commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1081,375 @@ def _n2(a, b):
+                if m < 0:
+                    ix = ~ix
+            elif m == mr and l.exp == r.exp and (
+                    l.exp.is_integer or
```
These conditions are too strict and are keeping the `_base_like_pow` lines unused.

smichr commented on sympy/core/assumptions.py:
```diff
@@ -385,6 +385,14 @@ def check_assumptions(expr, against=None, **assume):
 
+    To see if a number matches the assumptions of an expression, pass
+    the number as the first argument, else it's specific assumptions
+    may not have a non-None value is the expression:
```
```suggestion
    may not have a non-None value in the expression:
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1081,383 @@ def _n2(a, b):
+    is_ge(x, S.NegativeOne), is_ge(S.Zero, x)))
+
+# b**e >= 1
+_big_ppow = lambda b, e: fuzzy_or((
+        fuzzy_and((b.is_zero, e.is_zero)),  # 1
+        fuzzy_and((Mod(e, 2) == 0, fuzzy_or((
+        is_ge(S.NegativeOne, b) and is_ge(e, S.Zero),
+        _small_neg(b) and is_ge(S.Zero, e))))),
+    _small_pos(b) and is_ge(S.Zero, e),
+    is_ge(b, S.One) and is_ge(e, S.Zero)))
```
I still think these should be proper functions with docstrings and tests.

oscarbenjamin replied:
They don't need to be "public" but the meaning should be clearer in the code.

smichr replied:
Yes -- once this is pared down as much as possible I will add final touches.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1075,9 +1081,383 @@ def _n2(a, b):
+        if pause: input('press return to continue')
+
+    # work really hard to identify a power for comparison
+    if l.is_Pow and r == l.base:
+        r = Pow(r, 1, evaluate=False)
+    elif r.is_Pow and l == r.base:
+        l = Pow(l, 1, evaluate=False)
```
The idea behind introducing multiple dispatch here was supposed to be that we wouldn't just put everything in an `Expr,Expr` handler.

smichr replied:
Are we supposed to create a bunch of permutations of the types:

Add, Add
Add, Pow
Pow, Add
Pow, Pow
Add, Mul
Mul, Add
Pow, Mul
Mul, Pow

oscarbenjamin replied:
That was the intention. Maybe it wasn't the best design if that can't be made to work.

This was intended to be more or less a part of the old assumptions and the expectation was not that it would resolve all cases. Instead only "easy" cases would be resolved. The new assumptions are a more natural place to have a slower but more complete implementation for resolving inequalities like this.

smichr replied:
I'll give it a try to put it into the handlers. The good news is that already tests are running significantly faster with even these changes in place.

smichr committed:
test floor/ceiling args relative to +/-1

smichr committed:
enhance Relational comparisons

smichr committed:
wip-cov

smichr committed:
coverage

smichr committed:
Merge branch 'master' into unit-frac

smichr committed:
Merge branch 'master' into unit-frac
